File: src/comms/mqtt/MQTTSubscriber.cpp
#include "comms/mqtt/MQTTSubscriber.hpp"

XBT_LOG_NEW_DEFAULT_CATEGORY(mqtt_subscriber, "MQTT Subscriber");

namespace enigma {
namespace mqtt {

namespace sg4 = simgrid::s4u;

MQTTSubscriber::MQTTSubscriber(const std::string& broker, const std::string& sub_id)
    : broker_name(broker) {
    
    // Generate subscriber ID if not provided
    if (sub_id.empty()) {
        auto* host = sg4::this_actor::get_host();
        subscriber_id = std::string(host->get_cname()) + "_sub_" + 
                       std::to_string(sg4::this_actor::get_pid());
    } else {
        subscriber_id = sub_id;
    }
    
    broker_mbox = sg4::Mailbox::by_name(MQTTBroker::get_broker_mailbox(broker_name));
    my_mbox = sg4::Mailbox::by_name(subscriber_id);
    
    XBT_DEBUG("MQTT Subscriber '%s' initialized (broker: %s)",
              subscriber_id.c_str(), broker_name.c_str());
}

void MQTTSubscriber::subscribe(const std::string& topic) {
    auto* ctrl_msg = MQTTControlMessage::subscribe(topic, subscriber_id);
    
    XBT_DEBUG("Subscribing to topic '%s'", topic.c_str());
    
    broker_mbox->put(ctrl_msg, sizeof(MQTTControlMessage));
    subscribed_topics.push_back(topic);
}

void MQTTSubscriber::unsubscribe(const std::string& topic) {
    auto* ctrl_msg = MQTTControlMessage::unsubscribe(topic, subscriber_id);
    
    XBT_DEBUG("Unsubscribing from topic '%s'", topic.c_str());
    
    broker_mbox->put(ctrl_msg, sizeof(MQTTControlMessage));
    
    // Remove from local list
    auto it = std::find(subscribed_topics.begin(), subscribed_topics.end(), topic);
    if (it != subscribed_topics.end()) {
        subscribed_topics.erase(it);
    }
}

std::shared_ptr<MQTTMessage> MQTTSubscriber::receive(double timeout) {
    try {
        std::shared_ptr<MQTTMessage>* msg_ptr;
        
        if (timeout > 0) {
            msg_ptr = my_mbox->get<std::shared_ptr<MQTTMessage>>(timeout);
        } else {
            msg_ptr = my_mbox->get<std::shared_ptr<MQTTMessage>>();
        }
        
        auto msg = *msg_ptr;
        delete msg_ptr;
        
        XBT_DEBUG("Received message from topic '%s' (size: %zu bytes)",
                  msg->topic.c_str(), msg->size);
        
        return msg;
        
    } catch (const simgrid::TimeoutException&) {
        XBT_DEBUG("Receive timeout");
        return nullptr;
    }
}

bool MQTTSubscriber::has_messages() const {
    return my_mbox->ready();
}

const std::string& MQTTSubscriber::get_mailbox_name() const {
    return subscriber_id;
}

} // namespace mqtt
} // namespace enigma

File: src/python/enigma/comms/mqtt/mqtt_subscriber.py
"""
MQTTSubscriber: subscribes to topics and receives messages via the simulated broker.
"""
from __future__ import annotations

from typing import List, Optional
import simgrid
from .mqtt_broker import MQTTBroker, MQTTMessage, _ControlMessage, _ControlType


class MQTTSubscriber:
    """
    A subscriber registers itself with the broker over the control mailbox
    and subsequently receives messages from the broker via a per-subscriber
    delivery mailbox.

    Parameters
    ----------
    broker_name:
        Must match the name used when calling :func:`~.mqtt_broker.start_broker`.
    subscriber_id:
        Optional human-readable name; defaults to the current actor's host name
        combined with a random suffix to guarantee uniqueness.
    """

    def __init__(
        self,
        broker_name: str = "mqtt_broker",
        subscriber_id: str = "",
    ) -> None:
        self.broker_name = broker_name
        host_name = simgrid.this_actor.get_host().name
        self._subscriber_id = subscriber_id or f"{host_name}_sub_{id(self)}"
        self._broker_mbox = simgrid.Mailbox.by_name(
            MQTTBroker.get_broker_mailbox(broker_name)
        )
        # Unique delivery mailbox for this subscriber instance
        self._my_mbox = simgrid.Mailbox.by_name(self._delivery_mbox_name)
        self._subscribed_topics: List[str] = []

    # ------------------------------------------------------------------ #
    # Subscription management                                             #
    # ------------------------------------------------------------------ #

    def subscribe(self, topic: str) -> None:
        """Subscribe to *topic*; messages will be routed to this subscriber's mailbox."""
        if topic not in self._subscribed_topics:
            # Tell broker about the subscription using the *delivery* mailbox name
            ctrl = _ControlMessage(
                ctrl_type=_ControlType.SUBSCRIBE,
                topic=topic,
                subscriber=self._delivery_mbox_name,
            )
            self._broker_mbox.put(ctrl, 64)   # small control packet
            self._subscribed_topics.append(topic)

    def unsubscribe(self, topic: str) -> None:
        """Unsubscribe from *topic*."""
        if topic in self._subscribed_topics:
            ctrl = _ControlMessage(
                ctrl_type=_ControlType.UNSUBSCRIBE,
                topic=topic,
                subscriber=self._delivery_mbox_name,
            )
            self._broker_mbox.put(ctrl, 64)
            self._subscribed_topics.remove(topic)

    # ------------------------------------------------------------------ #
    # Receiving messages                                                  #
    # ------------------------------------------------------------------ #

    def receive(self, timeout: float = -1.0) -> Optional[MQTTMessage]:
        """
        Block until a message arrives or until *timeout* seconds elapse.

        Parameters
        ----------
        timeout:
            Maximum wait in simulated seconds.  Use ``-1`` for indefinite wait.

        Returns
        -------
        :class:`~.mqtt_broker.MQTTMessage` or ``None`` if timeout was reached.
        """
        try:
            if timeout < 0:
                return self._my_mbox.get()
            else:
                comm = self._my_mbox.get_async()
                comm.wait_for(timeout)
                return comm.get_payload()
        except (simgrid.TimeoutException, simgrid.NetworkFailureException):
            return None
        except Exception:
            return None

    def has_messages(self) -> bool:
        """Return True if there is at least one message waiting (non-blocking check)."""
        try:
            return self._my_mbox.ready
        except Exception:
            return False

    # ------------------------------------------------------------------ #
    # Properties                                                          #
    # ------------------------------------------------------------------ #

    @property
    def subscriber_id(self) -> str:
        return self._subscriber_id

    @property
    def mailbox_name(self) -> str:
        return self._delivery_mbox_name

    @property
    def subscribed_topics(self) -> List[str]:
        return list(self._subscribed_topics)

    @property
    def _delivery_mbox_name(self) -> str:
        return f"subscriber_{self._subscriber_id}"
